pkg/controllers/provisioningv2/rke2/machineprovision/controller.go

```diff
@@ -623,29 +646,33 @@ func constructFilesSecret(driver string, config map[string]interface{}) (*corev1
 	return nil, nil
 }
 
-func shouldForceRemoval(job *batchv1.Job, d data.Object) bool {
-	var createJobCondition summary.Condition
-	for _, c := range summary.GetUnstructuredConditions(d) {
-		if c.Type() == createJobConditionType {
-			createJobCondition = c
-			break
-		}
+func shouldCleanupObjects(job *batchv1.Job, d data.Object) bool {
+	if !job.Status.CompletionTime.IsZero() {
+		return true
+	}
+
+	createJobCondition := getCondition(d, createJobConditionType)
+	if createJobCondition == nil {
+		return false
 	}
 
 	forceRemoveAnnValue, _ := d.Map("metadata", "annotations")[forceRemoveMachineAnn].(string)
 
-	if createJobCondition.Reason() != string(errors.CreateMachineError) &&
-		strings.ToLower(forceRemoveAnnValue) != "true" {
-		return false
+	if createJobCondition.Reason() == string(errors.CreateMachineError) || strings.ToLower(forceRemoveAnnValue) == "true" {
+		return strings.ToLower(job.Spec.Template.Labels[InfraJobRemove]) == "true" && condition.Cond("Failed").IsTrue(job)
```

Comment:
This is changed because the previous implementation was incorrect (thanks Sergey).

Reply:
To clarify based on offline discussion - the previous implementation was actually correct and produces the same outputs for the same permutations of input conditions. However, the new implementation matches natural thinking more closely and thus is easier to digest.